File: packages/core/src/rules/block.ts
import type { Element } from 'domhandler';
import { isTag } from 'domhandler';
import type { Rule } from '../types.js';
import { getTextContent } from './util.js';

/**
 * Bare CSS class names on `<pre>` elements that should be treated as
 * code-block languages. Covers diagram-as-code tools that render into
 * `<pre class="toolname">` without a `language-` prefix.
 */
const BARE_LANG_CLASSES = new Set(['mermaid']);

/**
 * Detect the language identifier for a `<pre>` code block.
 *
 * Resolution order:
 * 1. `language-*` class on a `<code>` child element
 * 2. `language-*` class on the `<pre>` element itself (Prism.js pattern)
 * 3. A recognized bare class name on the `<pre>` element (e.g. `mermaid`)
 */
function detectLanguage(node: Element): string {
    const codeChild = node.children.find((c): c is Element => isTag(c) && c.name === 'code');

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- class may be undefined at runtime
    const codeLang = codeChild?.attribs.class?.match(/language-(\S+)/)?.[1];
    if (codeLang) return codeLang;

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- class may be undefined at runtime
    const preLang = node.attribs.class?.match(/language-(\S+)/)?.[1];
    if (preLang) return preLang;

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- class may be undefined at runtime
    for (const cls of node.attribs.class?.split(/\s+/) ?? []) {
        if (BARE_LANG_CLASSES.has(cls)) return cls;
    }

    return '';
}

/**
 * Built-in rules for block-level HTML elements.
 *
 * Handles headings (`h1`–`h6`), paragraphs, blockquotes, preformatted code blocks,
 * horizontal rules, line breaks, and stripped elements (`script`, `style`, `head`, etc.).
 * Headings support both ATX (`#`) and setext (underline) styles via the `headingStyle` option.
 */
export const blockRules: Rule[] = [
    {
        filter: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'],
        replacement: ({ node, convertChildren, options }) => {
            const level = Number.parseInt(node.name[1], 10);
            const content = convertChildren(node).trim();
            if (!content) return null;

            if (options.headingStyle === 'setext' && level <= 2) {
                const ch = level === 1 ? '=' : '-';
                return `\n\n${content}\n${ch.repeat(content.length)}\n\n`;
            }
            return `\n\n${'#'.repeat(level)} ${content}\n\n`;
        },
        priority: 0
    },

    {
        filter: 'p',
        replacement: ({ convertChildren, node }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            return `\n\n${content}\n\n`;
        },
        priority: 0
    },

    {
        filter: 'blockquote',
        replacement: ({ convertChildren, node }) => {
            const content = convertChildren(node).trim();
            if (!content) return null;
            const quoted = content
                .split('\n')
                .map(line => `> ${line}`)
                .join('\n');
            return `\n\n${quoted}\n\n`;
        },
        priority: 0
    },

    {
        filter: 'pre',
        replacement: ({ node, options }) => {
            const lang = detectLanguage(node);
            const text = getTextContent(node);
            const fence = options.fenceChar.repeat(3);
            return `\n\n${fence}${lang}\n${text}\n${fence}\n\n`;
        },
        priority: 0
    },

    {
        filter: 'hr',
        replacement: () => '\n\n---\n\n',
        priority: 0
    },

    {
        filter: 'br',
        replacement: () => '  \n',
        priority: 0
    },

    {
        filter: ['script', 'style', 'noscript', 'template'],
        replacement: () => null,
        priority: 0
    },

    // Strip document metadata elements — <head>, <title>, etc. contain
    // metadata that should never appear as visible Markdown content.
    {
        filter: ['head', 'title', 'meta', 'link', 'base'],
        replacement: () => null,
        priority: 0
    }
];

File: python/src/markdown_for_agents/rules/block.py
from __future__ import annotations

import re

from .._types import Rule, RuleContext
from ..dom.nodes import Element, is_tag
from .util import get_text_content

_BARE_LANG_CLASSES = frozenset({"mermaid"})


def _detect_language(node: Element) -> str:
    """Detect the language identifier for a <pre> code block."""
    # Check <code> child element
    code_child: Element | None = None
    for c in node.children:
        if is_tag(c) and isinstance(c, Element) and c.name == "code":
            code_child = c
            break

    if code_child is not None:
        cls = code_child.attribs.get("class", "")
        match = re.search(r"language-(\S+)", cls)
        if match:
            return match.group(1)

    # Check <pre> element itself
    pre_cls = node.attribs.get("class", "")
    match = re.search(r"language-(\S+)", pre_cls)
    if match:
        return match.group(1)

    # Check bare class names
    for cls in pre_cls.split():
        if cls in _BARE_LANG_CLASSES:
            return cls

    return ""


block_rules: list[Rule] = [
    Rule(
        filter=["h1", "h2", "h3", "h4", "h5", "h6"],
        replacement=lambda ctx: _heading_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="p",
        replacement=lambda ctx: _p_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="blockquote",
        replacement=lambda ctx: _blockquote_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="pre",
        replacement=lambda ctx: _pre_replacement(ctx),
        priority=0,
    ),
    Rule(
        filter="hr",
        replacement=lambda ctx: "\n\n---\n\n",
        priority=0,
    ),
    Rule(
        filter="br",
        replacement=lambda ctx: "  \n",
        priority=0,
    ),
    Rule(
        filter=["script", "style", "noscript", "template"],
        replacement=lambda ctx: None,
        priority=0,
    ),
    Rule(
        filter=["head", "title", "meta", "link", "base"],
        replacement=lambda ctx: None,
        priority=0,
    ),
]


def _heading_replacement(ctx: RuleContext) -> str | None:
    level = int(ctx.node.name[1])
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None

    if ctx.options.heading_style == "setext" and level <= 2:
        ch = "=" if level == 1 else "-"
        return f"\n\n{content}\n{ch * len(content)}\n\n"
    return f"\n\n{'#' * level} {content}\n\n"


def _p_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    return f"\n\n{content}\n\n"


def _blockquote_replacement(ctx: RuleContext) -> str | None:
    content = ctx.convert_children(ctx.node).strip()
    if not content:
        return None
    quoted = "\n".join(f"> {line}" for line in content.split("\n"))
    return f"\n\n{quoted}\n\n"


def _pre_replacement(ctx: RuleContext) -> str:
    lang = _detect_language(ctx.node)
    text = get_text_content(ctx.node)
    fence = ctx.options.fence_char * 3
    return f"\n\n{fence}{lang}\n{text}\n{fence}\n\n"
